### PR TITLE
Launch superweapons using warhead

### PR DESCRIPTION
**Pull request by chaserli**
Reopened the PR for nightly builds for testers

### Launch superweapons on impact

- Superweapons can be launched when a standard warhead is detonated. This means that animation weapons from _Ares_ are **not** supported.(even when working together with Starkku's animation damage re-implementation) Nevertheless, animation warheads may works in certain circumstances. Due to the nature of some superweapon types, not all superweapons are suitable for launch. The user should not abuse this feature.
  - `LaunchSW` specifies the superweapons to launch when the warhead is detonated.
  - `LaunchSW.RealLaunch` controls whether the owner who fired the warhead must own all listed superweapons and sufficient fund to support `Money.Amout`. Otherwise they will be launched out of nowhere.
  - `LaunchSW.IgnoreInhibitors` ignores `SW.Inhibitors` of the superweapon, otherwise only non-inhibited superweapons are launched.

In `rulesmd.ini`:
```ini
[SOMEWARHEAD]                 ; Warhead
LaunchSW=                     ; list of superweapons
LaunchSW.RealLaunch=no        ; bool
LaunchSW.IgnoreInhibitors=no  ; bool

### CODE REVIEW

**Comment by 8abs** [comment]
This PR is super useful. Hoping it could be merged into phobos build version.

File: docs/locale/zh_CN/LC_MESSAGES/New-or-Enhanced-Logics.po [hunk 1]
```diff
@@ -2727,6 +2727,33 @@ msgid ""
...
+"the warhead must own all listed superweapons and sufficient fund to "
+"support `Money.Amout`. Otherwise they will be launched out of nowhere."
+msgstr "`LaunchSW.RealLaunch` 真实发射指使用引爆弹头所属方目前现有的超武。否则超武将无条件发射到指定格子，（无视是否拥有，是否够钱，是否充能）"
+
+#: ../New-or-Enhanced-Logics.md:568
+msgid ""
+"`LaunchSW.IgnoreInhibitors` ignores `SW.Inhibitors` of the superweapon,"
+" otherwise only non-inhibited superweapons are launched. "
+msgstr "`LaunchSW.IgnoreInhibitors` 是否忽略`SW.Inhibitors`"
+
```
Comment (Metadorius) [inline docs/locale/zh_CN/LC_MESSAGES/New-or-Enhanced-Logics.po:2756]: Would need updates after the doc fixes.

File: src/Ext/SWType/Body.cpp [hunk 2]
```diff
@@ -60,6 +62,10 @@ void SWTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI) {
 
 	this->LimboKill_Affected.Read(exINI, pSection, "LimboKill.Affected");
 	this->LimboKill_IDs.Read(exINI, pSection, "LimboKill.IDs");
+	
+	// inhibitor related
+	this->SW_Inhibitors.Read(exINI, pSection, "SW.Inhibitors");
+	this->SW_AnyInhibitor.Read(exINI, pSection, "SW.AnyInhibitor");
```
Comment (Metadorius) [inline src/Ext/SWType/Body.cpp:68]: If you're parsing the tags in the end, then you should declare and serialize also in the end, so the order is the same, just for the readability sake.

File: src/Ext/SWType/Body.cpp [hunk 3]
```diff
@@ -13,6 +13,8 @@ template <typename T>
 void SWTypeExt::ExtData::Serialize(T& Stm) {
 	Stm
 		.Process(this->Money_Amount)
+		.Process(this->SW_Inhibitors)
+		.Process(this->SW_AnyInhibitor)
```
Comment (Metadorius) [inline src/Ext/SWType/Body.cpp:17]: See next comment.

File: src/Ext/SWType/Body.h [hunk 4]
```diff
@@ -24,10 +23,17 @@ class SWTypeExt
...
+		Valueable<int> Money_Amount;
+		ValueableVector<TechnoTypeClass*> SW_Inhibitors;
+		Valueable<bool> SW_AnyInhibitor;
+
 		ValueableVector<ValueableVector<int>> LimboDelivery_RandomWeightsData;
 
 		ExtData(SuperWeaponTypeClass* OwnerObject) : Extension<SuperWeaponTypeClass>(OwnerObject)
 			, Money_Amount { 0 }
+			, SW_Inhibitors {}
+			, SW_AnyInhibitor{ false }
```
Comment (Metadorius) [inline src/Ext/SWType/Body.h:36]: See prev. comment.

File: src/Ext/SWType/FireSuperWeapon.cpp [hunk 5]
```diff
@@ -191,3 +191,58 @@ void SWTypeExt::ExtData::ApplyLimboKill(HouseClass* pHouse)
...
+
+bool SWTypeExt::HasInhibitor(SWTypeExt::ExtData* pSWType, HouseClass* pOwner, const CellStruct& Coords)
+{
+	// does not allow inhibitors
+	if (pSWType->SW_Inhibitors.empty() && !pSWType->SW_AnyInhibitor)
+		return false;
+
+	// a single inhibitor in range suffices
+	return std::any_of(TechnoClass::Array->begin(), TechnoClass::Array->end(), [=, &Coords](TechnoClass* pTechno)
+						{ return IsInhibitorEligible(pSWType, pOwner, Coords, pTechno);}
```
Comment (Metadorius) [inline src/Ext/SWType/FireSuperWeapon.cpp:246]: Not sure if this (one-line and wrong indentation) is OK.

**Review by Metadorius — CHANGES_REQUESTED** [review]
Looks good except few code style issues, good to merge after fixing.

**Comment by chaserli** [comment]
My local branch falls behind develop, I don't know how to resolve these conflicts

**Review by Metadorius — APPROVED** [review]

**Comment by Metadorius** [comment]
image

**Comment by Metadorius** [comment]
Fixed

**Pull request merged**